### PR TITLE
Add send gcode check

### PR DESCRIPTION
**Pull request by Subcircuit**
### Description

Add check and error response according to the [reprap.org](https://reprap.org/wiki/G-code#N:_Line_number) description.
The line number will be cleared before choosing to print the file.

Include N line number  , * checksum.
e.g. N123 [...G Code in here...] *123

### Related Issues

~~If two errors appear in the same line number, the error content will be sent.~~

### CODE REVIEW

File: TFT/src/User/API/interfaceCmd.c [hunk 1]
```diff
@@ -888,6 +902,30 @@ void sendQueueCmd(void)
...
+    char line_str[CMD_MAX_CHAR]={0};
+
+    infoCmd.line++;
+    sprintf(line_str, "N%u ", infoCmd.line);
+    strcat(line_str, infoCmd.queue[infoCmd.index_r].gcode);
+
+    while(line_str[index] != '\n') index++;
+
+    num = index;
+    for(; index; )
```
Comment (traffic-light) [inline TFT/src/User/API/interfaceCmd.c:920]: Please use curly braces. Even when it's one line

File: TFT/src/User/API/interfaceCmd.c [hunk 2]
```diff
@@ -888,6 +902,30 @@ void sendQueueCmd(void)
...
+      cs = cs ^ line_str[--index];
+    cs &= 0xff;
+    sprintf(line_str+num, "*%d\n", cs);
+    strcpy(infoCmd.queue[infoCmd.index_r].gcode, line_str);
+  }
+  if(checking_flag==0)
```
Comment (traffic-light) [inline TFT/src/User/API/interfaceCmd.c:926]: Please use curly braces. Even when it's one line

File: TFT/src/User/API/interfaceCmd.c [hunk 3]
```diff
@@ -888,6 +902,30 @@ void sendQueueCmd(void)
...
+    checking_flag = 1;
+#endif //GCODE_CHECKING
   Serial_Puts(SERIAL_PORT, infoCmd.queue[infoCmd.index_r].gcode);
   if (avoid_terminal != true){
```
Comment (traffic-light) [inline TFT/src/User/API/interfaceCmd.c:930]: Please place the curly braces on the next line

**Review by traffic-light — COMMENTED** [review]
Please use the curly braces where needed and place them on the second line.

And can you please do a space after if: `if (mycheck == true)`

**Comment by Subcircuit** [comment]
> 请在需要的地方使用花括号并将其放在第二行。
> 
> 而且，如果发生以下情况`if (mycheck == true)`
> ，您可以在下面留一个空格吗：还用于while循环，for循环等

Thanks for your reminding

**Comment by Subcircuit** [comment]
>Related Issues
>If two errors appear in the same line number, the error content will be sent.

Fix issues  (e38e873)


> Please use the curly braces where needed and place them on the second line.
> 
> And can you please do a space after if: `if (mycheck == true)`
> Also for while loops, for loops and so on

@traffic-light  Thanks for your reminding (98cb911)

**Comment by traffic-light** [comment]
@Subcircuit no problemo!

**Comment by oldman4U** [comment]
Before I test this, is there any notification which informs the user about an error?

I believe it would also be good to attach some examples so there is a way to test the functionality.

My 2 cents

**Comment by Subcircuit** [comment]
@oldman4U  
If the transmission is wrong, pop-up display error message, and automatically resends the error line.  
For example the first line number error shows "Line Number is not Last Line Number+1, Last Line: 0".  
If the checksum error is "checksum mismatch, Last Line: 0".

**Comment by oldman4U** [comment]
Hi. 

I believe this is a cool feature but how can I get a "broken" gcode file?

**Comment by Codel1417** [comment]
sounds like either sending a broken check by ommitting the *123 portion would be the right way but im unsure if check and retry needs to be initialized first.

**Comment by Codel1417** [comment]
```
int cs = 0; for(i = 0; cmd[i] != '*' && cmd[i] != NULL; i++) cs = cs ^ cmd[i]; cs &= 0xff; // Defensive programming...
```
This would have to be implemented in another program to generate a checksum for testing, or just send a broken check with a random number

**Comment by Codel1417** [comment]
N1 G28 *65 should fail for testing

**Comment by Domush** [comment]
RepRap requires checksums to be calculated prior to inline gCode comments. This code would require a check for comments after the gCode command and calculate&insert the checksum before the comment.

Additionally, a check is needed for existing N## (although rare) in SD gcode in order to avoid an error.

**Pull request closed without merging**